File: packages/components/src/image.ts
/**
 * Image组件 - 对齐Flutter Image API
 * 用于显示图片，支持网络图片、本地资源、文件等
 */

import * as React from 'react';
import type { ColorValue } from './styles/color';
import type { BoxFitValue, ImageRepeatValue, BlendModeValue, AlignmentValue } from './styles/types';

/**
 * 图片源类型
 */
export type ImageSourceType = 'network' | 'asset' | 'file' | 'memory';

/**
 * Image组件属性
 */
export interface ImageProps {
  /** 图片源URL或路径（必需） */
  src: string;
  
  /** 图片源类型 */
  srcType?: ImageSourceType;
  
  // 尺寸属性
  /** 图片宽度 */
  width?: number;
  /** 图片高度 */
  height?: number;
  
  // 适配属性
  /** 图片适配方式 */
  fit?: BoxFitValue;
  /** 图片对齐方式 */
  alignment?: AlignmentValue;
  /** 图片重复方式 */
  repeat?: ImageRepeatValue;
  
  // 颜色滤镜
  /** 颜色滤镜 */
  color?: ColorValue;
  /** 颜色混合模式 */
  colorBlendMode?: BlendModeValue;
  
  // 性能优化
  /** 缓存宽度 */
  cacheWidth?: number;
  /** 缓存高度 */
  cacheHeight?: number;
  
  // 无障碍
  /** 语义标签 */
  semanticLabel?: string;
  
  /** 唯一标识 */
  id?: string;
}

/**
 * Image组件
 * 对应Flutter的Image widget
 * 
 * @example
 * // 网络图片
 * <Image 
 *   src="https://example.com/image.png" 
 *   srcType="network"
 *   width={300}
 *   height={200}
 *   fit="cover"
 * />
 * 
 * @example
 * // 本地资源
 * <Image 
 *   src="assets/images/logo.png" 
 *   srcType="asset"
 *   width={100}
 *   height={100}
 * />
 * 
 * @example
 * // 带颜色滤镜
 * <Image 
 *   src="https://example.com/image.png"
 *   color={Color.fromRGBO(255, 0, 0, 0.5)}
 *   colorBlendMode="multiply"
 *   fit="contain"
 * />
 */
export const Image: React.FC<ImageProps> = ({
  src,
  srcType = 'network', // 默认为网络图片
  width,
  height,
  fit,
  alignment,
  repeat,
  color,
  colorBlendMode,
  cacheWidth,
  cacheHeight,
  semanticLabel,
  id,
}) => {
  return React.createElement('Image', {
    src,
    srcType,
    width,
    height,
    fit,
    alignment,
    repeat,
    color,
    colorBlendMode,
    cacheWidth,
    cacheHeight,
    semanticLabel,
    id,
  });
};

File: lib/components/widgets/image.dart
import 'dart:io';
import 'package:flutter/material.dart';
import '../base/flutter_component.dart';
import '../../core/virtual_dom_parser.dart';
import '../../utils/style_parsers/foundation/primitive_parsers.dart';
import '../../utils/style_parsers/foundation/geometry_parsers.dart';

/// Flutter Image组件映射
/// 
/// 对应React端的Image组件
/// 支持网络图片、本地资源、文件图片等
class ImageComponent extends FlutterComponent {
  @override
  String get componentType => 'Image';

  @override
  String get description => 'Flutter Image组件映射';

  @override
  Map<String, Type> get supportedProps => {
    'src': String,
    'srcType': String,
    'width': num,
    'height': num,
    'fit': String,
    'alignment': String,
    'repeat': String,
    'color': dynamic,
    'colorBlendMode': String,
    'cacheWidth': int,
    'cacheHeight': int,
    'semanticLabel': String,
    'id': String,
  };

  @override
  bool get supportsChildren => false;

  @override
  Widget build(VirtualDOM vdom) {
    final src = vdom.getProp<String>('src');
    
    if (src == null || src.isEmpty) {
      throw FlutterError('Image组件必须设置src属性');
    }

    final srcType = vdom.getProp<String>('srcType') ?? 'network';
    final width = vdom.getProp<num>('width')?.toDouble();
    final height = vdom.getProp<num>('height')?.toDouble();
    final fit = _parseBoxFit(vdom.getProp<String>('fit'));
    final alignment = GeometryParsers.parseAlignment(vdom.getProp<String>('alignment')) ?? Alignment.center;
    final repeat = _parseImageRepeat(vdom.getProp<String>('repeat')) ?? ImageRepeat.noRepeat;
    final color = PrimitiveParsers.parseColor(vdom.getProp('color'));
    final colorBlendMode = _parseBlendMode(vdom.getProp<String>('colorBlendMode'));
    final cacheWidth = vdom.getProp<int>('cacheWidth');
    final cacheHeight = vdom.getProp<int>('cacheHeight');
    final semanticLabel = vdom.getProp<String>('semanticLabel');
    final id = vdom.getProp<String>('id');

    // 根据srcType选择对应的Image构造函数
    // 直接在Image组件上设置key，而不是用KeyedSubtree包裹
    switch (srcType) {
      case 'network':
        return Image.network(
          src,
          key: id != null ? Key(id) : null,
          width: width,
          height: height,
          fit: fit,
          alignment: alignment,
          repeat: repeat,
          color: color,
          colorBlendMode: colorBlendMode,
          cacheWidth: cacheWidth,
          cacheHeight: cacheHeight,
          semanticLabel: semanticLabel,
        );
        
      case 'asset':
        return Image.asset(
          src,
          key: id != null ? Key(id) : null,
          width: width,
          height: height,
          fit: fit,
          alignment: alignment,
          repeat: repeat,
          color: color,
          colorBlendMode: colorBlendMode,
          cacheWidth: cacheWidth,
          cacheHeight: cacheHeight,
          semanticLabel: semanticLabel,
        );
        
      case 'file':
        return Image.file(
          File(src),
          key: id != null ? Key(id) : null,
          width: width,
          height: height,
          fit: fit,
          alignment: alignment,
          repeat: repeat,
          color: color,
          colorBlendMode: colorBlendMode,
          cacheWidth: cacheWidth,
          cacheHeight: cacheHeight,
          semanticLabel: semanticLabel,
        );
        
      default:
        // 默认使用network
        return Image.network(
          src,
          key: id != null ? Key(id) : null,
          width: width,
          height: height,
          fit: fit,
          alignment: alignment,
          repeat: repeat,
          color: color,
          colorBlendMode: colorBlendMode,
          cacheWidth: cacheWidth,
          cacheHeight: cacheHeight,
          semanticLabel: semanticLabel,
        );
    }
  }

  /// 解析BoxFit
  BoxFit? _parseBoxFit(String? value) {
    if (value == null) return null;
    
    switch (value) {
      case 'fill':
        return BoxFit.fill;
      case 'contain':
        return BoxFit.contain;
      case 'cover':
        return BoxFit.cover;
      case 'fitWidth':
        return BoxFit.fitWidth;
      case 'fitHeight':
        return BoxFit.fitHeight;
      case 'none':
        return BoxFit.none;
      case 'scaleDown':
        return BoxFit.scaleDown;
      default:
        return null;
    }
  }

  /// 解析ImageRepeat
  ImageRepeat? _parseImageRepeat(String? value) {
    if (value == null) return null;
    
    switch (value) {
      case 'repeat':
        return ImageRepeat.repeat;
      case 'repeatX':
        return ImageRepeat.repeatX;
      case 'repeatY':
        return ImageRepeat.repeatY;
      case 'noRepeat':
        return ImageRepeat.noRepeat;
      default:
        return null;
    }
  }

  /// 解析BlendMode
  BlendMode? _parseBlendMode(String? value) {
    if (value == null) return null;
    
    switch (value) {
      case 'clear':
        return BlendMode.clear;
      case 'src':
        return BlendMode.src;
      case 'dst':
        return BlendMode.dst;
      case 'srcOver':
        return BlendMode.srcOver;
      case 'dstOver':
        return BlendMode.dstOver;
      case 'srcIn':
        return BlendMode.srcIn;
      case 'dstIn':
        return BlendMode.dstIn;
      case 'srcOut':
        return BlendMode.srcOut;
      case 'dstOut':
        return BlendMode.dstOut;
      case 'srcATop':
        return BlendMode.srcATop;
      case 'dstATop':
        return BlendMode.dstATop;
      case 'xor':
        return BlendMode.xor;
      case 'plus':
        return BlendMode.plus;
      case 'modulate':
        return BlendMode.modulate;
      case 'screen':
        return BlendMode.screen;
      case 'overlay':
        return BlendMode.overlay;
      case 'darken':
        return BlendMode.darken;
      case 'lighten':
        return BlendMode.lighten;
      case 'colorDodge':
        return BlendMode.colorDodge;
      case 'colorBurn':
        return BlendMode.colorBurn;
      case 'hardLight':
        return BlendMode.hardLight;
      case 'softLight':
        return BlendMode.softLight;
      case 'difference':
        return BlendMode.difference;
      case 'exclusion':
        return BlendMode.exclusion;
      case 'multiply':
        return BlendMode.multiply;
      case 'hue':
        return BlendMode.hue;
      case 'saturation':
        return BlendMode.saturation;
      case 'color':
        return BlendMode.color;
      case 'luminosity':
        return BlendMode.luminosity;
      default:
        return null;
    }
  }
}
